openquake/calculators/event_based.py

```diff
@@ -241,35 +240,51 @@ def save_ruptures(self, ruptures_by_grp_id):
                     self.datastore.extend(
                         ev, numpy.array(events, calc.stored_event_dt))
 
-                # save rup_data
-                if hasattr(ruptures_by_grp_id, 'rup_data'):
-                    trt = self.grp_trt[grp_id]
-                    sm_id = self.sm_by_grp[grp_id]
-                    key = 'rup_data/sm-%04d/%s' % (sm_id, trt)
-                    self.rup_data[trt] = self.datastore.extend(
-                            key, ruptures_by_grp_id.rup_data)
+            # save rup_data
+            if hasattr(ruptures_by_grp_id, 'rup_data'):
+                for grp_id, data in sorted(
+                        ruptures_by_grp_id.rup_data.items()):
+                    key = 'rup_data/grp-%02d' % grp_id
+                    self.rup_data = self.datastore.extend(key, data)
 
     def post_execute(self, result):
         """
         Save the SES collection
         """
+        logging.info('Setting event years')
+        inv_time = int(self.oqparam.investigation_time)
+        numpy.random.seed(self.oqparam.random_seed)
+        for sm in self.datastore['events']:
+            set_random_years(self.datastore['events/' + sm].value, inv_time)
+
         nr = sum_dict(result)
         logging.info('Saved %d ruptures, %d events',
                      nr, sum(self.eid.values()))
         if 'ruptures' in self.datastore:
             self.datastore.set_nbytes('ruptures')
         self.datastore.set_nbytes('events')
 
-        for dset in self.rup_data.values():
+        for dset in self.datastore['rup_data'].values():
             if len(dset):
                 numsites = dset['numsites']
                 multiplicity = dset['multiplicity']
                 spr = numpy.average(numsites, weights=multiplicity)
                 mul = numpy.average(multiplicity, weights=numsites)
                 self.datastore.set_attrs(dset.name, sites_per_rupture=spr,
                                          multiplicity=mul)
-        if self.rup_data:
-            self.datastore.set_nbytes('rup_data')
+        self.datastore.set_nbytes('rup_data')
+
+
+def set_random_years(events, investigation_time):
+    """
+    Sort the `events` array and attach year labels sensitive to the
+    SES ordinal and the investigation time.
+    """
+    events.sort(order='rupserial')
+    years = numpy.random.choice(investigation_time, len(events))
```

Comment:
It would be better to have the first year labeled as year 1 rather than as year 0.